### PR TITLE
Features branch Private Events

### PR DESCRIPTION
**Pull request by tigamadou**
- User Authentication 
  - The user can register
  - The user can sign in and sign out
  - The user has many events
- Event Management
  - A user can create many events
  - A user can attend to many events
  - An event can be attended by many users.
  - An event belongs to a user
- Tests with Rspec and Capybara
  - unit tests for models associations and validations
  - integrations tests for authentication and events management
 - Code linter.

### CODE REVIEW

File: app/models/event.rb [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+  # def self.upcoming
+  #     Event.all.where("date > ?", Time.now)
+  # end
+
+  scope :upcoming, -> { where('date > ?', Time.now) }
+  scope :past, -> { where('date < ?', Time.now) }
+
+  scope :upcoming_events, -> { where('date > ?', Time.now) }
+  scope :previous_events, -> { where('date < ?', Time.now) }
+end
```
Comment (maelfosso) [inline app/models/event.rb]: Why those two? Why did you repeat them?

File: app/models/event.rb [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+  has_many :event_attendees
+  has_many :attendees, class_name: 'User', through: :event_attendees
+  validates :name, :description, :date, :venue, presence: true
+  # def self.past
+  #     Event.all.where("date < ?", Time.now)
+  # end
+
+  # def self.upcoming
+  #     Event.all.where("date > ?", Time.now)
+  # end
```
Comment (maelfosso) [inline app/models/event.rb]: Remove the comments, they are not allowed.

File: app/models/user.rb [hunk 3]
```diff
@@ -0,0 +1,8 @@
+class User < ApplicationRecord
+  validates :username, :email, :password, presence: true
+
+  has_secure_password
+  has_many :event
```
Comment (maelfosso) [inline app/models/user.rb]: It's preferable to write `:events`

File: app/models/user.rb [hunk 4]
```diff
@@ -0,0 +1,8 @@
...
+  has_many :event_attendees
+  has_many :attended_event, class_name: 'Event', through: :event_attendees
```
Comment (maelfosso) [inline app/models/user.rb]: it's preferable to use `:attended_events`

File: app/controllers/user_controller.rb [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+      redirect_to user_path(@user)
+    else
+      render 'new'
+    end
+  end
+
+  def show
+    @user = User.find(params[:id])
+    @prev_events = current_user.attended_event.previous_events
+    @upcoming_events = current_user.attended_event.upcoming_events
```
Comment (maelfosso) [inline app/controllers/user_controller.rb]: Not for the current user but for the user which the `id` in the `params` that is `@user`

**Review by maelfosso — CHANGES_REQUESTED** [review]
# Changes required :red_circle: 

Hello,
Thank you for that well-done project.
However, there are some changes to do before move forward.

1. Please, no need to use `HomeController` and the layout `layout/home`. Please, use **`ApplicationController`** and **`layout/application.html.erb`**

2. `User#show` is not well working. Please, read the comment below.

3. I got this error when trying to create an account.
Screenshot from 2020-03-12 23-15-44

Kindly make the requested changes and submit a new code review request.
Thanks

Mael FOSSO

File: app/models/event.rb (hunk 1 above)
Comment (tigamadou) [reply app/models/event.rb]: Duplicate removed.

File: app/models/event.rb (hunk 2 above)
Comment (tigamadou) [reply app/models/event.rb]: Removed

File: app/models/user.rb (hunk 3 above)
Comment (tigamadou) [reply app/models/user.rb]: Done.

File: app/models/user.rb (hunk 4 above)
Comment (tigamadou) [reply app/models/user.rb]: Done

File: app/controllers/user_controller.rb (hunk 5 above)
Comment (tigamadou) [reply app/controllers/user_controller.rb]: Done.

**Review by Torres-ssf — APPROVED** [review]
### Project Approved :+1: :tada: 
Hi there team :hand: 
Nice job implementing the requested changes :smiley:
I must say, the UI is looking really good :clap: 
**I'm approving this project**
Good luck with the next one.

[OPTIONAL]
Is there a way to a user attend to an event through the UI?
I manage to attend to event's through rails console but I didn't see any way to do it in the UI. I think this will be a nice feature for your app to have.

Happy Coding
**Sergio Torres**

**Comment by tigamadou** [comment]
Thank you @Torres-ssf 😄 
I think I'll implement that feature too. 👍

**Pull request merged**